File: JavaScript/4-Stack/Problems/4-LargestRectangle-in-Histogram/LargestRectangle-in-HistogramPro.js
/*
Largest Rectangle in Histogram: JavaScript Solutions

Problem:
Given an array of integers heights representing the histogram's bar heights where the width of each bar is 1, find the area of the largest rectangle in the histogram.

Constraints:
* 1 <= heights.length <= 10^5
* 0 <= heights[i] <= 10^4
*/

/**
 * Approach 1: Brute Force
 *
 * For each bar, consider it as the minimum height and expand left and right.
 * Calculate the area with this minimum height and update the maximum area.
 *
 * Time Complexity: O(n^2)
 * Space Complexity: O(1)
 */
function largestRectangleAreaBruteForce(heights) {
    let maxArea = 0;
    const n = heights.length;

    for (let i = 0; i < n; i++) {
        let minHeight = heights[i];
        for (let j = i; j < n; j++) {
            minHeight = Math.min(minHeight, heights[j]);
            const width = j - i + 1;
            maxArea = Math.max(maxArea, minHeight * width);
        }
    }
    return maxArea;
}

/**
 * Approach 2: Divide and Conquer
 *
 * Recursively divide the histogram into subproblems.
 * The largest rectangle can be in the left subproblem, right subproblem,
 * or across the middle.
 *
 * Time Complexity: O(n log n) in average case, O(n^2) in worst case
 * Space Complexity: O(n log n)
 */
function largestRectangleAreaDivideAndConquer(heights, start, end) {
    if (start > end) {
        return 0;
    }

    if (start === end) {
        return heights[start];
    }

    let minIndex = start;
    for (let i = start + 1; i <= end; i++) {
        if (heights[i] < heights[minIndex]) {
            minIndex = i;
        }
    }

    const leftArea = largestRectangleAreaDivideAndConquer(heights, start, minIndex - 1);
    const rightArea = largestRectangleAreaDivideAndConquer(heights, minIndex + 1, end);
    const middleArea = heights[minIndex] * (end - start + 1);

    return Math.max(leftArea, rightArea, middleArea);
}

function largestRectangleAreaDivideAndConquerMain(heights) {
    return largestRectangleAreaDivideAndConquer(heights, 0, heights.length - 1);
}


/**
 * Approach 3: Stack
 *
 * Use a stack to keep track of increasing heights.
 * When a smaller height is encountered, calculate the area with the heights in the stack.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
function largestRectangleAreaStack(heights) {
    const n = heights.length;
    const stack = [];
    let maxArea = 0;
    let i = 0;

    while (i < n || stack.length > 0) {
        if (i < n && (stack.length === 0 || heights[i] >= heights[stack[stack.length - 1]])) {
            stack.push(i);
            i++;
        } else {
            const top = stack.pop();
            const width = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;
            const area = heights[top] * width;
            maxArea = Math.max(maxArea, area);
        }
    }
    return maxArea;
}

/**
 * Approach 4: Optimized Stack
 *
 * Similar to the previous stack approach, but adds 0 to the beginning and end
 * of the heights array to simplify the boundary conditions.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
function largestRectangleAreaOptimizedStack(heights) {
    const n = heights.length;
    const modifiedHeights = [0, ...heights, 0]; // Add 0 at the beginning and end
    const stack = [];
    let maxArea = 0;

    for (let i = 0; i < modifiedHeights.length; i++) {
        while (stack.length > 0 && modifiedHeights[i] < modifiedHeights[stack[stack.length - 1]]) {
            const top = stack.pop();
            const width = i - stack[stack.length - 1] - 1;
            const area = modifiedHeights[top] * width;
            maxArea = Math.max(maxArea, area);
        }
        stack.push(i);
    }
    return maxArea;
}

/**
 * Approach 5: Monolithic Stack (Most Efficient)
 *
 * This approach uses a single pass and a stack to maintain indices of monotonically increasing heights.  It's the most efficient
 * and commonly used solution in real-world applications due to its optimal performance.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
function largestRectangleAreaMonolithicStack(heights) {
    const n = heights.length;
    const stack = []; // Stack to store *indices* of bars, maintaining increasing height order.
    let maxArea = 0;

    for (let i = 0; i <= n; i++) {
        // Iterate through the heights, including a sentinel (virtual bar of height 0) at the end (i = n).
        // The sentinel ensures that all bars in the stack are processed.
        const currentHeight = i === n ? 0 : heights[i];

        // While the stack is not empty and the current height is *less* than the height of the bar at the
        // index at the top of the stack, we've found a potential right boundary for the rectangle
        // formed by the bar at the top of the stack.
        while (stack.length > 0 && currentHeight < heights[stack[stack.length - 1]]) {
            const topIndex = stack.pop(); // Get the index of the bar at the top of the stack.
            const barHeight = heights[topIndex]; // Height of the popped bar.

            // Calculate the width of the rectangle.
            // If the stack is empty, it means the popped bar extends to the beginning of the histogram.
            // Otherwise, the width is the distance between the current bar (i) and the bar at the
            // new top of the stack (exclusive).
            const width = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;

            // Calculate the area of the rectangle.
            const area = barHeight * width;
            maxArea = Math.max(maxArea, area); // Update the maximum area found so far.
        }

        // Push the *index* of the current bar onto the stack.  This maintains the monotonically increasing
        // property of the stack.  If the current height is greater than or equal to the height of the bar
        // at the top of the stack, it means we can potentially expand the rectangle formed by the
        // current bar to the right.
        stack.push(i);
    }

    return maxArea;
}



// Example Usage and Verification
const heights = [2, 1, 5, 6, 2, 3];
console.log("Heights:", heights);

// Brute Force
const result1 = largestRectangleAreaBruteForce(heights);
console.log("Largest Rectangle Area (Brute Force):", result1); // Expected: 10

// Divide and Conquer
const result2 = largestRectangleAreaDivideAndConquerMain(heights);
console.log("Largest Rectangle Area (Divide and Conquer):", result2); // Expected: 10

// Stack
const result3 = largestRectangleAreaStack(heights);
console.log("Largest Rectangle Area (Stack):", result3); // Expected: 10

// Optimized Stack
const result4 = largestRectangleAreaOptimizedStack(heights);
console.log("Largest Rectangle Area (Optimized Stack):", result4); // Expected: 10

// Monolithic Stack
const result5 = largestRectangleAreaMonolithicStack(heights);
console.log("Largest Rectangle Area (Monolithic Stack):", result5); // Expected: 10

File: Java/4-Stack/Problems/4-LargestRectangle-in-Histogram/LargestRectangle-in-HistogramPro.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.stream.IntStream;
import java.util.Arrays;
import java.util.function.IntFunction;

public class LargestRectangleInHistogram {

    /**
     * Approach 1: Brute Force
     *
     * Description:
     * The most straightforward approach.  For each bar in the histogram, we consider it as the minimum height
     * of a potential rectangle and expand left and right as far as possible while the heights are greater
     * or equal to the current bar's height.  We calculate the area for each such rectangle and keep track of the maximum.
     *
     * Time Complexity: O(n^2), where n is the number of bars in the histogram.
     * Space Complexity: O(1)
     *
     * Real-world application:
     * -   Useful for a very small dataset or as a baseline for comparing the performance of more optimized algorithms.
     * -   Can be used in simple data analysis scripts where performance is not critical.
     *
     * @param heights The array of bar heights in the histogram.
     * @return The largest rectangular area in the histogram.
     */
    public int largestRectangleAreaBruteForce(int[] heights) {
        int maxArea = 0;
        int n = heights.length;

        for (int i = 0; i < n; i++) {
            int minHeight = heights[i];
            for (int j = i; j < n; j++) {
                minHeight = Math.min(minHeight, heights[j]); // Find the minimum height in the current range
                maxArea = Math.max(maxArea, minHeight * (j - i + 1)); // Calculate area and update maxArea
            }
        }
        return maxArea;
    }

    /**
     * Approach 2: Divide and Conquer
     *
     * Description:
     * A divide-and-conquer approach.  We find the minimum height bar in the histogram.  The largest rectangle
     * must either:
     * 1.  Include the minimum height bar.
     * 2.  Be entirely to the left of the minimum height bar.
     * 3.  Be entirely to the right of the minimum height bar.
     * We recursively calculate the largest rectangle for the left and right sub-histograms and combine the results.
     *
     * Time Complexity: O(n log n) in the average case, O(n^2) in the worst case (when the array is sorted).
     * Space Complexity: O(n) due to the recursion stack.
     *
     * Real-world application:
     * -   Useful in scenarios where the input data can be effectively divided into subproblems, such as
     * parallel processing of large datasets.
     * -   Can be adapted for distributed computing environments.
     *
     * @param heights The array of bar heights in the histogram.
     * @return The largest rectangular area in the histogram.
     */
    public int largestRectangleAreaDivideAndConquer(int[] heights) {
        if (heights == null || heights.length == 0) {
            return 0;
        }
        return divideAndConquer(heights, 0, heights.length - 1);
    }

    private int divideAndConquer(int[] heights, int left, int right) {
        if (left > right) {
            return 0;
        }

        int minIndex = findMinIndex(heights, left, right); // Find index of the minimum height bar
        int currentArea = heights[minIndex] * (right - left + 1); // Area with the minimum bar

        // Recursively find the largest area in the left and right sub-histograms
        int leftArea = divideAndConquer(heights, left, minIndex - 1);
        int rightArea = divideAndConquer(heights, minIndex + 1, right);

        return Math.max(currentArea, Math.max(leftArea, rightArea)); // Return the maximum of the three areas
    }

    private int findMinIndex(int[] heights, int left, int right) {
        int minIndex = left;
        for (int i = left + 1; i <= right; i++) {
            if (heights[i] < heights[minIndex]) {
                minIndex = i;
            }
        }
        return minIndex;
    }

    /**
     * Approach 3: Stack
     *
     * Description:
     * The most efficient approach.  We use a stack to keep track of the indices of the bars.  The stack maintains the
     * property that the heights of the bars corresponding to the indices in the stack are increasing.
     * When we encounter a bar whose height is less than the height of the bar at the top of the stack, we pop the stack
     * and calculate the area of the rectangle with the popped bar as the minimum height.
     *
     * Time Complexity: O(n), where n is the number of bars in the histogram.
     * Space Complexity: O(n) for the stack.
     *
     * Real-world application:
     * -   This approach is highly optimized and suitable for real-time data processing, such as analyzing stock prices
     * or monitoring system performance metrics where identifying the largest rectangular area within a dataset
     * is crucial.
     * -   Used in applications requiring efficient processing of sequential data.
     *
     * @param heights The array of bar heights in the histogram.
     * @return The largest rectangular area in the histogram.
     */
    public int largestRectangleAreaStack(int[] heights) {
        int maxArea = 0;
        int n = heights.length;
        Deque<Integer> stack = new ArrayDeque<>(); // Use ArrayDeque as a stack (more efficient)
        int i = 0;

        while (i < n || !stack.isEmpty()) { // Iterate until all bars are processed and stack is empty.
            if (i < n && (stack.isEmpty() || heights[i] >= heights[stack.peek()])) {
                stack.push(i++); // Push the current bar's index onto the stack
            } else {
                int top = stack.pop(); // Pop the top of the stack
                // Calculate area with heights[top] as the smallest bar.
                int area = heights[top] * (stack.isEmpty() ? i : i - stack.peek() - 1);
                maxArea = Math.max(maxArea, area); // Update maxArea
            }
        }
        return maxArea;
    }

    /**
     * Approach 4: Stack (Optimized)
     *
     * Description:
     * An optimized version of the stack approach.  To avoid the extra check `i < n` in the loop,
     * we pad the input `heights` array with 0 at the beginning and end. This ensures that all bars
     * are eventually popped from the stack.
     *
     * Time Complexity: O(n), where n is the number of bars in the histogram.
     * Space Complexity: O(n) for the stack.
     *
     * Real-world application:
     * -   This optimized approach is suitable for high-performance applications where minimizing
     * micro-optimizations can lead to noticeable performance gains, such as in high-frequency trading
     * platforms or real-time data analysis systems.
     * -   Useful when dealing with a very large number of histograms.
     *
     * @param heights The array of bar heights in the histogram.
     * @return The largest rectangular area in the histogram.
     */
    public int largestRectangleAreaStackOptimized(int[] heights) {
        int maxArea = 0;
        int n = heights.length;
        int[] paddedHeights = new int[n + 2]; // Create a padded array
        System.arraycopy(heights, 0, paddedHeights, 1, n); // Copy original heights to the middle
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(0); // Push 0 at the beginning

        for (int i = 1; i <= n + 1; i++) { // Iterate through the padded array
            while (paddedHeights[i] < paddedHeights[stack.peek()]) {
                int top = stack.pop();
                int area = paddedHeights[top] * (i - stack.peek() - 1);
                maxArea = Math.max(maxArea, area);
            }
            stack.push(i);
        }
        return maxArea;
    }

    /**
     * Approach 5: Using Streams (Java 8 and above) - Not ideal for performance critical applications.
     *
     * Description:
     * This approach uses Java 8 streams for a more functional style.  It iterates through each bar and calculates the maximum
     * possible area with that bar as the minimum height.  While concise, this approach is generally less efficient than
     * the stack-based methods due to the overhead of stream operations.
     *
     * Time Complexity: O(n^2) in most practical scenarios due to the nested loops within the stream operations.
     * Space Complexity: O(n) in the worst case, due to the creation of streams and collections.
     *
     * Real-world application:
     * -   Useful for quick prototyping or in applications where code readability and conciseness are prioritized
     * over raw performance.  For example, in data analysis scripts or reporting tools.
     * -   Good for educational purposes to demonstrate functional programming concepts in Java.
     *
     * @param heights The array of bar heights in the histogram.
     * @return The largest rectangular area in the histogram.
     */
    public int largestRectangleAreaStreams(int[] heights) {
        if (heights == null || heights.length == 0) {
            return 0;
        }

        final int n = heights.length; // Make n final
        return IntStream.range(0, n)
                .map(i -> {
                    int minHeight = heights[i];
                    int area = IntStream.range(i, n)
                            .map(j -> {
                                // Calculate minHeight within the lambda
                                int currentMinHeight = heights[i];
                                for (int k = i; k <= j; k++) {
                                    currentMinHeight = Math.min(currentMinHeight, heights[k]);
                                }
                                return currentMinHeight * (j - i + 1);
                            })
                            .max()
                            .orElse(0);
                    return area;
                })
                .max()
                .orElse(0);
    }

    public static void main(String[] args) {
        int[] heights = {2, 1, 5, 6, 2, 3};
        LargestRectangleInHistogram calculator = new LargestRectangleInHistogram();

        System.out.println("Heights: " + Arrays.toString(heights));
        System.out.println("Largest Rectangle Area (Brute Force): " + calculator.largestRectangleAreaBruteForce(heights));
        System.out.println("Largest Rectangle Area (Divide and Conquer): " + calculator.largestRectangleAreaDivideAndConquer(heights));
        System.out.println("Largest Rectangle Area (Stack): " + calculator.largestRectangleAreaStack(heights));
        System.out.println("Largest Rectangle Area (Stack Optimized): " + calculator.largestRectangleAreaStackOptimized(heights));
        System.out.println("Largest Rectangle Area (Streams): " + calculator.largestRectangleAreaStreams(heights));

        // Example with another dataset
        int[] heights2 = {4, 2, 0, 3, 2, 5};
        System.out.println("\nHeights: " + Arrays.toString(heights2));
        System.out.println("Largest Rectangle Area (Brute Force): " + calculator.largestRectangleAreaBruteForce(heights2));
        System.out.println("Largest Rectangle Area (Divide and Conquer): " + calculator.largestRectangleAreaDivideAndConquer(heights2));
        System.out.println("Largest Rectangle Area (Stack): " + calculator.largestRectangleAreaStack(heights2));
        System.out.println("Largest Rectangle Area (Stack Optimized): " + calculator.largestRectangleAreaStackOptimized(heights2));
        System.out.println("Largest Rectangle Area (Streams): " + calculator.largestRectangleAreaStreams(heights2));

        // Example with monotonically increasing heights
        int[] heights3 = {1, 2, 3, 4, 5};
        System.out.println("\nHeights: " + Arrays.toString(heights3));
        System.out.println("Largest Rectangle Area (Brute Force): " + calculator.largestRectangleAreaBruteForce(heights3));
        System.out.println("Largest Rectangle Area (Divide and Conquer): " + calculator.largestRectangleAreaDivideAndConquer(heights3));
        System.out.println("Largest Rectangle Area (Stack): " + calculator.largestRectangleAreaStack(heights3));
        System.out.println("Largest Rectangle Area (Stack Optimized): " + calculator.largestRectangleAreaStackOptimized(heights3));
        System.out.println("Largest Rectangle Area (Streams): " + calculator.largestRectangleAreaStreams(heights3));

        // Example with monotonically decreasing heights
        int[] heights4 = {5, 4, 3, 2, 1};
        System.out.println("\nHeights: " + Arrays.toString(heights4));
        System.out.println("Largest Rectangle Area (Brute Force): " + calculator.largestRectangleAreaBruteForce(heights4));
        System.out.println("Largest Rectangle Area (Divide and Conquer): " + calculator.largestRectangleAreaDivideAndConquer(heights4));
        System.out.println("Largest Rectangle Area (Stack): " + calculator.largestRectangleAreaStack(heights4));
        System.out.println("Largest Rectangle Area (Stack Optimized): " + calculator.largestRectangleAreaStackOptimized(heights4));
        System.out.println("Largest Rectangle Area (Streams): " + calculator.largestRectangleAreaStreams(heights4));

        // Example with repeated heights
        int[] heights5 = {2, 2, 2, 2, 2};
        System.out.println("\nHeights: " + Arrays.toString(heights5));
        System.out.println("Largest Rectangle Area (Brute Force): " + calculator.largestRectangleAreaBruteForce(heights5));
        System.out.println("Largest Rectangle Area (Divide and Conquer): " + calculator.largestRectangleAreaDivideAndConquer(heights5));
        System.out.println("Largest Rectangle Area (Stack): " + calculator.largestRectangleAreaStack(heights5));
        System.out.println("Largest Rectangle Area (Stack Optimized): " + calculator.largestRectangleAreaStackOptimized(heights5));
        System.out.println("Largest Rectangle Area (Streams): " + calculator.largestRectangleAreaStreams(heights5));
    }
}

